### PR TITLE
[5.3] Optimize Arr methods

### PR DESCRIPTION
**Pull request by bepsvpt**
- Arr::exists

Using isset to optimize the process, only if the value is null will use array_key_exists to determine the given key exists.
- Arr::last

Same as #15213, prevent array copy.
- Arr::forget, Arr::set

Use foreach to prevent multiple array_shift.

### CODE REVIEW

File: src/Illuminate/Support/Arr.php [hunk 1]
```diff
@@ -120,7 +120,7 @@ public static function exists($array, $key)
             return $array->offsetExists($key);
         }
 
-        return array_key_exists($key, $array);
+        return isset($array[$key]) || array_key_exists($key, $array);
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Arr.php]: array_key_exists is pretty much the same speed on php 7, so you've actually made this slower

Comment (phanan) [inline src/Illuminate/Support/Arr.php]: …and even if it's slower, the difference is [super micro](https://ilia.ws/archives/247-Performance-Analysis-of-isset-vs-array_key_exists.html) and just not worth it if I may add. Unless, of course, you're calling this method repeatedly for like 10,000 times.

Comment (GrahamCampbell) [inline src/Illuminate/Support/Arr.php]: Even then, you're making this twice as slow for cases where it is there and is null.

Comment (fernandobandeira) [inline src/Illuminate/Support/Arr.php]: There's a benchmark example in [php.net](http://php.net/manual/en/function.array-key-exists.php)

> Benchmark (100000 runs):
array_key_exists() : 205 ms
isset() : 35ms
isset() || array_key_exists() : 48ms

However if you are running this 100000 times i guess that 150ms is not your main concern.

**Comment by taylorotwell** [comment]
Holding off on this for now because there are no benchmarks to prove the optimizations are relevant.

File: src/Illuminate/Support/Arr.php (hunk 1 above)
Comment (BePsvPT) [inline src/Illuminate/Support/Arr.php]: I only test this on local machine, and adding isset is faster.

However, when using sandbox online, array_key_exists is faster.

Benchmark : http://sandbox.onlinephpfunctions.com/code/90069414dc1b46c666fbfd96e1b9415b9d151378

**Comment by BePsvPT** [comment]
Benchmarks:

Arr::forget: http://sandbox.onlinephpfunctions.com/code/161e998b5eadba07d9b76feb61c629e8142176d6

Arr::set: http://sandbox.onlinephpfunctions.com/code/324e9cdb0d9c6662ceb30a1db534fd1e0fd206e9

forget, set are original and _forget,  _set are optimized

**Pull request closed without merging**